### PR TITLE
minimal GPUArray integration

### PR DESCRIPTION
**Pull request by SimonDanisch**

### CODE REVIEW

**Comment by SimonDanisch** [comment]
bump

**Comment by SimonDanisch** [comment]
bump - any comments?

**Comment by SimonDanisch** [comment]
bump

**Comment by MikeInnes** [comment]
Pretty busy but will try to take a look this week. I need to go over all the mechanisms you're buying CuArrays into here, e.g. the functionality the abstract type gives me and the generic kernel mechanisms.

**Comment by SimonDanisch** [comment]
bump! Would be nice to get this done sooner than later, since we already get into duplicated work for bug fixing - and I don't think I can fully announce GPUArrays without this.
It shouldn't take that much time, since 50% of there kernels are already from CuArrays and there are only like ~20. I'm willing to also adapt the broadcast a bit, since you seem to overload the Base interface more gracefully.

**Comment by MikeInnes** [comment]
Sorry for the wait. The BLR gpu machine is down right now but I'll definitely pick out some time for this next week. I might still want to make the patch a bit smaller and take things piece by piece, but this PR will be a really useful starting point for that.

One thing I've been thinking about is whether it's actually right to have a `GPUArray` abstract type here, or if that's the wrong way to express that. The problem is that an abstract type can only express properties along one dimension, and Base has chosen to give properties of the data layout (e.g. `DenseArray`), so using it for GPU-ness prevents us from working with that. On top of which, there's no way to dispatch on `CPUArray` right now in a way that's consistent. An alternative might be to have something like a trait, which defaults to CPU but can also be GPU, distributed etc.

Just a thought, doesn't have to block this but seems worth discussing.

**Comment by MikeInnes** [comment]
I've avoided implementing `pointer` so far because it seems easy to make mistakes with; people almost always use it to mean "cpu pointer" and it might be safer to have a separate `GPUArrays.pointer` function that's specific to GPU memory. Any thoughts?

File: src/blas/highlevel.jl [hunk 1]
```diff
@@ -10,8 +10,17 @@ CublasArray{T<:CublasFloat} = CuArray{T}
 #
 ###########
 
-Base.scale!(x::CuArray{T}, k::Number) where T<:CublasFloat =
-  scal!(length(x), convert(eltype(x), k), x, 1)
+function Base.scale!(X::CuArray{T}, s::Real) where T <: Base.BLAS.BlasComplex
```
Comment (MikeInnes) [inline src/blas/highlevel.jl]: Why is this needed? cublas already has methods for complex numbers

Comment (SimonDanisch) [reply src/blas/highlevel.jl]: Ambiguities! But I will double check.

File: src/gpuarray_interface.jl [hunk 2]
```diff
@@ -0,0 +1,110 @@
+import GPUArrays
+
+Base.similar(::Type{<: CuArray}, ::Type{T}, size::Base.Dims{N}) where {T, N} = CuArray{T, N}(size)
```
Comment (MikeInnes) [inline src/gpuarray_interface.jl]: This doesn't seem to have an analogue for other array types?

File: src/gpuarray_interface.jl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+#Abstract GPU interface
+immutable CUKernelState end
```
Comment (MikeInnes) [inline src/gpuarray_interface.jl]: should probably use `Cu` consistently

File: test/runtests.jl [hunk 4]
```diff
@@ -21,7 +21,29 @@ function testf(f, xs...)
   @test collect(f(cu.(xs)...)) ≈ collect(f(xs...))
 end
 
+
+using Base.Test, GPUArrays.TestSuite
+using GPUArrays: JLArray
```
Comment (MikeInnes) [inline test/runtests.jl]: What's this for?

**Comment by MikeInnes** [comment]
RE types, given that GPUArray is a DenseArray, what's your plan for when we have support for strided and sparse arrays that aren't dense?

File: src/gpuarray_interface.jl (hunk 2 above)
Comment (SimonDanisch) [reply src/gpuarray_interface.jl]: When you want to generically offer constructors like CuArray(rand(10)) you only have a type and no instance.
This solves that problem ;) If it was up to me, I'd love to have this in Base being the only method to overload!

**Comment by SimonDanisch** [comment]
I haven't thought much about it, but I've always imagined them as a wrapper type, which doesn't inherit necessarily from GPUArrays:
```
 # AbstractSparseArray or whatever GPU equivalent we come up with in a sparse GPU Array package
struct SpareseArray <: AbstractSparseArray
values::GPUArray
indices::GPUArray
end
```

**Comment by MikeInnes** [comment]
Doesn't that defeat the point of having a `GPUArray` type that you can dispatch on, write kernels for etc? You go back to needing a trait, but then you would have two inconsistent mechanisms for doing the same thing.

**Comment by SimonDanisch** [comment]
True! Well, we can always add a node to the type hieararchy. I'd love to have a trait based approach, but currently we just don't have very nice tools for that.
Trait like approaches still take me back to c where you need to pass the length of a vector as a separate argument, making it impossible to treat an array as a single object ;)

File: src/blas/highlevel.jl (hunk 1 above)
Comment (MikeInnes) [reply src/blas/highlevel.jl]: Even with the ambiguities it makes sense to use CUBLAS's method for complex numbers rather than reimplementing it.

Comment (MikeInnes) [reply src/blas/highlevel.jl]: The ambiguity here is caused by GPUArrays wrapping the method for Real rather than Number. So I have to explicitly override that in order to opt-out I guess, which is a bit nasty.

Comment (SimonDanisch) [reply src/blas/highlevel.jl]: Okay lets fix that then ;)

**Comment by MikeInnes** [comment]
I still have concerns about this. It's heavily specialised for one specific kind of array, and it's not clear to me that it's going to make it easier, rather than harder, to share code once we start having new or custom array types on the gpu. The one thing it does provide – the `GPUArray` interface – is not really reliable since it won't work with constructs like `RowVector`, and it seems pretty short-sighted to have no plan for sparse or strided arrays.

It also still bothers me that it's not clear what this buys CuArrays into; you say there's some 20 kernels it adds, but there's no good way to verify that other than weeding through 3,000 lines of GPUArrays.jl code. As we discussed, an opt-in approach like the test suite has would easily resolve this and other issues. As it is I hope you're going to be around to resolve any issues that do come up.

For now I guess we'll just have to try it and see what happens.

File: src/blas/highlevel.jl (hunk 1 above)
Comment (SimonDanisch) [reply src/blas/highlevel.jl]: https://github.com/JuliaGPU/GPUArrays.jl/pull/83

**Comment by SimonDanisch** [comment]
Thanks mike! Well, it's the first design iteration, and I tried just to concentrate on sharing code between cuda and opencl for dense buffers. I hope to get you on board and in the future design things together! My main hope is, that we can now solve most problems inside GPUArrays so that improvements come free to all specialized array packages! We should get issues started for strided/sparse arrrays to figure out what would be needed for a better design! I think we also need a better story for e.g. adapt, views and device arrays!

**Comment by SimonDanisch** [comment]
Would it help if I port the nicer broadcast from CuArrays to GPUArrays?
With promote_container_type & generated kernel, while keeping the faster gpu_sub2ind...
I could do this today, but then we should move the implementation out of CuArrays ;)

**Comment by MikeInnes** [comment]
Sure, that seems like a good next step.
